java/ql/lib/semmle/code/java/ControlFlowGraph.qll

```diff
@@ -84,45 +84,122 @@ private import Completion
+
+  /** A node in the expression-level control-flow graph. */
+  class Node extends TNode {
+    /** Gets the statement containing this node, if any. */
+    Stmt getEnclosingStmt() {
+      result = this.asStmt() or
+      result = this.asExpr().getEnclosingStmt()
+    }
+
+    /** Gets the immediately enclosing callable whose body contains this node. */
+    Callable getEnclosingCallable() {
+      this = TExitNode(result) or
+      result = this.asStmt().getEnclosingCallable() or
+      result = this.asExpr().getEnclosingCallable()
+    }
+
+    /** Gets the statement this `Node` corresponds to, if any. */
+    Stmt asStmt() { this = TStmtNode(result) }
+
+    /** Gets the expression this `Node` corresponds to, if any. */
+    Expr asExpr() { this = TExprNode(result) }
+
+    /** Gets the call this `Node` corresponds to, if any. */
+    Call asCall() {
+      result = this.asExpr() or
+      result = this.asStmt()
+    }
 
-  /** Gets an immediate successor of this node. */
-  ControlFlowNode getASuccessor() { result = succ(this) }
+    /** Gets an immediate successor of this node. */
+    Node getASuccessor() { result = succ(this) }
 
-  /** Gets an immediate predecessor of this node. */
-  ControlFlowNode getAPredecessor() { this = succ(result) }
+    /** Gets an immediate predecessor of this node. */
+    Node getAPredecessor() { this = succ(result) }
 
-  /** Gets an exception successor of this node. */
-  ControlFlowNode getAnExceptionSuccessor() { result = succ(this, ThrowCompletion(_)) }
+    /** Gets an exception successor of this node. */
+    Node getAnExceptionSuccessor() { result = succ(this, ThrowCompletion(_)) }
+
+    /** Gets a successor of this node that is neither an exception successor nor a jump (break, continue, return). */
+    Node getANormalSuccessor() {
+      result = succ(this, BooleanCompletion(_, _)) or
+      result = succ(this, NormalCompletion())
+    }
+
+    /** Gets the basic block that contains this node. */
+    BasicBlock getBasicBlock() { result.getANode() = this }
+
+    /** Gets a textual representation of this element. */
+    string toString() {
+      result = this.asExpr().toString()
+      or
+      result = this.asStmt().toString()
+      or
+      result = "Exit" and this instanceof ExitNode
+    }
+
+    /** Gets the source location for this element. */
+    Location getLocation() {
+      result = this.asExpr().getLocation() or
+      result = this.asStmt().getLocation() or
+      result = this.(ExitNode).getEnclosingCallable().getLocation()
+    }
 
-  /** Gets a successor of this node that is neither an exception successor nor a jump (break, continue, return). */
-  ControlFlowNode getANormalSuccessor() {
-    result = succ(this, BooleanCompletion(_, _)) or
-    result = succ(this, NormalCompletion())
+    /**
+     * Get the most appropriate AST node for this control flow node, if any.
+     *
+     * This is needed for the equivalence relation on basic blocks in range
+     * analysis.
```

Comment:
That's not its only use. Also, it seems reasonable to have predicate like this without such a specific-sounding constraint.
```suggestion
     *
     * This is needed for the equivalence relation on basic blocks in range
     * analysis.
```

Reply:
Did you mean to suggest a different wording?

Reply:
Ah no, I meant to simply delete it. Edited.

Reply:
Would you expect every control flow node to be able to supply this? Or is it optional?

Reply:
Hmm, there are two different potentially relevant predicates, one that provides the inverse of the `getControlFlowNode` on `Expr` and `Stmt`, and one that projects every Cfg node to an associated Ast node. I think I'd name the latter something with `project`, and let this predicate take the semantics of the former. Hence, I'd say that this is optional and that we should postfix its qldoc with the customary "if any".